grpc/interceptors.go

```diff
@@ -421,3 +424,90 @@ func (dd deadlineDetails) Error() string {
 	return fmt.Sprintf("%s.%s timed out after %d ms",
 		dd.service, dd.method, int64(dd.latency/time.Millisecond))
 }
+
+// serviceAuthChecker provides two server interceptors which can check that
+// every request for a given gRPC service is being made over an mTLS connection
+// from a client which is allow-listed for that particular service.
+type serviceAuthChecker struct {
+	serviceClientNames map[string]map[string]struct{}
+}
+
+// newServiceAuthChecker takes a GRPCServerConfig and uses its Service stanzas
+// to construct a serviceAuthChecker which enforces the service/client mappings
+// contained in the config.
+func newServiceAuthChecker(c *cmd.GRPCServerConfig) *serviceAuthChecker {
+	names := make(map[string]map[string]struct{})
+	for serviceName, service := range c.Services {
+		names[serviceName] = make(map[string]struct{})
+		for _, clientName := range service.ClientNames {
+			names[serviceName][clientName] = struct{}{}
+		}
+	}
+	return &serviceAuthChecker{names}
+}
+
+// UnaryInterceptor returns a gRPC unary interceptor.
+func (ac *serviceAuthChecker) UnaryInterceptor() func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		err := ac.checkContextAuth(ctx, info.FullMethod)
+		if err != nil {
+			return nil, err
+		}
+		return handler(ctx, req)
+	}
+}
+
+// StreamInterceptor returns a gRPC stream interceptor.
+func (ac *serviceAuthChecker) StreamInterceptor() func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		err := ac.checkContextAuth(ss.Context(), info.FullMethod)
+		if err != nil {
+			return err
+		}
+		return handler(srv, ss)
+	}
+}
+
+// checkContextAuth does most of the heavy lifting. It extracts TLS information
+// from the incoming context, gets the set of DNS names contained in the client
+// mTLS cert, and returns nil if at least one of those names appears in the set
+// of allowed client names for given service (or if the set of allowed client
+// names is empty).
+func (ac *serviceAuthChecker) checkContextAuth(ctx context.Context, fullMethod string) error {
+	serviceName, _ := splitMethodName(fullMethod)
+
+	allowedClientNames, ok := ac.serviceClientNames[serviceName]
+	if !ok || len(allowedClientNames) == 0 {
+		// Don't restrict clients if no allowlist has been configured.
+		return nil
+	}
```

Comment:
This took a surprising amount of refactoring, to be able to support "only conditionally add this interceptor" in a clean way that lets us assure correct ordering of interceptors regardless of whether they're disabled or not. But I think the result is an improvement overall. Done.

Reply:
Actually wait, this is more complex than I thought.

For example: the Remote VAs have one allowed client for their VA service (the VA itself) but zero allowed clients for their CAA service (because we only do single-perspective CAA). So the allowed clients list for that service is empty. And in the current architecture, we can't just say "don't register the CAA service on remote VAs", because remote VAs don't actually know that they're remote. There's no config value that unambiguously indicates that.